File: ios/Classes/Methods/ParticipantsMethods.swift
import Flutter
import TwilioConversationsClient

class ParticipantsMethods {
    public static func getUsers(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let conversationSid = arguments["conversationSid"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing conversationSid", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                var users = [TCHUser?]()
                conversation.participants().forEach({ participant in
                    participant.subscribedUser() { (result, user) in
                        users.append(user)
                        if users.count == conversation.participants().count {
                            let userList = users.compactMap { Mapper.userToDict($0) }
                            flutterResult(Mapper.encode(userList))
                        }
                    }
                })
            } else {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
    
    public static func getParticipantsList(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let conversationSid = arguments["conversationSid"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing conversationSid", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                let participantsMap = conversation.participants().map { Mapper.participantToDict($0, conversationSid: conversationSid)}
                flutterResult(Mapper.encode(participantsMap))
            } else {
                SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
    
    public static func addParticipantByIdentity(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let identity = arguments["identity"] as? String,
              let conversationSid = arguments["conversationSid"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                conversation.addParticipant(byIdentity: identity,
                                            attributes: nil) { (result: TCHResult) in
                    if result.isSuccessful {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                        flutterResult(true)
                    } else {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error adding participant to conversation with sid '\(conversationSid)', Error: \(result.error.debugDescription)", details: nil))
                    }
                }
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
    
    public static func removeParticipantByIdentity(_ call: FlutterMethodCall, _ flutterResult: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?],
              let identity = arguments["identity"] as? String,
              let conversationSid = arguments["conversationSid"] as? String else {
            return flutterResult(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }
        
        SwiftTwilioConversationsPlugin.instance?.client?.conversation(withSidOrUniqueName: conversationSid, completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                conversation.removeParticipant(byIdentity: identity) { (result: TCHResult) in
                    if result.isSuccessful {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onSuccess")
                        flutterResult(true)
                    } else {
                        SwiftTwilioConversationsPlugin.debug("\(call.method) => onError: \(String(describing: result.error))")
                        flutterResult(FlutterError(code: "ERROR", message: "Error removing participant from conversation with sid '\(conversationSid)', Error: \(result.error.debugDescription)", details: nil))
                    }
                }
            } else {
                flutterResult(FlutterError(code: "ERROR", message: "Error retrieving conversation with sid '\(conversationSid)'", details: nil))
            }
        })
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/ParticipantsMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.google.gson.Gson
import com.twilio.conversations.*
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin

object ParticipantsMethods {
    fun addParticipantByIdentity(call: MethodCall, result: MethodChannel.Result) {
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        try {
            TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
                override fun onSuccess(conversation: Conversation) {
                    conversation.addParticipantByIdentity(identity, Attributes(), object : StatusListener {
                        override fun onSuccess() {
                            TwilioConversationsPlugin.debug("${call.method} => onSuccess")
                            result.success(true)
                        }

                        override fun onError(errorInfo: ErrorInfo) {
                            TwilioConversationsPlugin.debug("${call.method} (conversation.addParticipantByIdentity) => onError: $errorInfo")
                            result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                        }
                    })
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }

            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }

    fun removeParticipantByIdentity(call: MethodCall, result: MethodChannel.Result) {
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)
        val identity = call.argument<String>("identity")
                ?: return result.error("ERROR", "Missing 'identity'", null)

        try {
            TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
                override fun onSuccess(conversation: Conversation) {
                    conversation.removeParticipantByIdentity(identity, object : StatusListener {
                        override fun onSuccess() {
                            TwilioConversationsPlugin.debug("${call.method} => onSuccess")
                            result.success(true)
                        }

                        override fun onError(errorInfo: ErrorInfo) {
                            TwilioConversationsPlugin.debug("${call.method} (conversation.removeParticipantByIdentity) => onError: $errorInfo")
                            result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                        }
                    })
                }

                override fun onError(errorInfo: ErrorInfo) {
                    TwilioConversationsPlugin.debug("${call.method} => onError: $errorInfo")
                    result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                }

            })
        } catch (err: IllegalArgumentException) {
            return result.error("IllegalArgumentException", err.message, null)
        }
    }


    fun getParticipantsList(call: MethodCall, result: MethodChannel.Result) {
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)

        TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
            override fun onSuccess(conversation: Conversation) {
                TwilioConversationsPlugin.debug("ParticipantsMethods.getParticipantsList (Conversation.getConversation) => onSuccess")
                val participantsListMap = Gson().toJson(Mapper.participantListToMap(conversation.participantsList))
                result.success(participantsListMap)
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioConversationsPlugin.debug("${call.method} onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun getUsers(call: MethodCall, result: MethodChannel.Result) {
        val conversationSid = call.argument<String>("conversationSid")
                ?: return result.error("ERROR", "Missing 'conversationSid'", null)

        TwilioConversationsPlugin.client?.getConversation(conversationSid, object : CallbackListener<Conversation> {
            override fun onSuccess(conversation: Conversation) {
                TwilioConversationsPlugin.debug("${call.method} => onSuccess")
                val users = mutableListOf<User>()
                conversation.participantsList.forEach {
                    it.getAndSubscribeUser { user ->
                        users.add(user)
                        if (users.size == conversation.participantsList.size) {
                            val userListMap = Gson().toJson(users.mapNotNull { Mapper.userToMap(it) })
                            result.success(userListMap)
                        }
                    }
                }
            }

            override fun onError(errorInfo: ErrorInfo) {
                TwilioConversationsPlugin.debug("${call.method} onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }
}